lib/manager/jenkins/extract.ts

```diff
@@ -1,33 +1,110 @@
+import yaml from 'js-yaml';
 import * as datasourceJenkins from '../../datasource/jenkins-plugins';
 import { logger } from '../../logger';
 import { SkipReason } from '../../types';
 import { isSkipComment } from '../../util/ignore';
 import * as dockerVersioning from '../../versioning/docker';
 import type { PackageDependency, PackageFile } from '../types';
+import type { JenkinsPlugin, JenkinsPlugins } from './types';
 
-export function extractPackageFile(content: string): PackageFile | null {
-  logger.trace('jenkins.extractPackageFile()');
+const YamlExtension = /\.ya?ml$/;
+
+function getDependency(plugin: JenkinsPlugin): PackageDependency {
+  const dep: PackageDependency = {
+    datasource: datasourceJenkins.id,
+    versioning: dockerVersioning.id,
+    depName: plugin.artifactId,
+  };
+
+  if (plugin.source?.version) {
+    dep.currentValue = plugin.source.version.toString();
+    if (typeof plugin.source.version !== 'string') {
+      dep.skipReason = SkipReason.InvalidVersion;
+      logger.warn(
+        { dep },
+        'Jenkins plugin dependency version is not a string and will be ignored'
+      );
+    }
+  } else {
+    dep.skipReason = SkipReason.NoVersion;
+  }
+
+  if (
+    plugin.source?.version === 'latest' ||
+    plugin.source?.version === 'experimental' ||
+    plugin.groupId
+  ) {
+    dep.skipReason = SkipReason.UnsupportedVersion;
+  }
+
+  if (plugin.source?.url) {
+    dep.skipReason = SkipReason.InternalPackage;
+  }
+
+  if (!dep.skipReason && plugin.renovate?.ignore) {
+    dep.skipReason = SkipReason.Ignored;
+  }
+
+  logger.debug({ dep }, 'Jenkins plugin dependency');
+  return dep;
+}
+
+function extractYaml(content: string): PackageDependency[] {
+  const deps: PackageDependency[] = [];
+
+  try {
+    const doc = yaml.safeLoad(content, { json: true }) as JenkinsPlugins;
+    if (doc?.plugins) {
+      for (const plugin of doc.plugins) {
+        if (plugin.artifactId) {
+          const dep = getDependency(plugin);
+          deps.push(dep);
+        }
+      }
+    }
+  } catch (err) {
+    logger.warn({ err }, 'Error parsing Jenkins plugins');
+    return null;
```

Comment:
needs coverage, maybe wrong test?

Reply:
The test with `invalid.yaml` use safeLoad and will not throw any exception. I am not sure how to force an exception with a mock and I don't see any examples.

Should I just re-add the code coverage ignore statement ? However, I can think we should keep the test.

Reply:
Checkout the recent gitlabci manager changes. I've a throwing test yaml there. 😏